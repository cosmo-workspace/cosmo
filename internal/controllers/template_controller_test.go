package controllers

import (
	"context"
	"time"

	. "github.com/cosmo-workspace/cosmo/pkg/snap"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	corev1 "k8s.io/api/core/v1"
	apierrs "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
	"sigs.k8s.io/controller-runtime/pkg/client"

	cosmov1alpha1 "github.com/cosmo-workspace/cosmo/api/v1alpha1"
	"github.com/cosmo-workspace/cosmo/pkg/instance"
)

var _ = Describe("Template controller", func() {
	const nsName string = "default"

	tmpl := cosmov1alpha1.Template{
		ObjectMeta: metav1.ObjectMeta{
			Name: "test-pod-tmpl1",
		},
		Spec: cosmov1alpha1.TemplateSpec{
			RawYaml: `apiVersion: v1
kind: Pod
metadata:
  name: alpine
spec:
  containers:
  - image: 'alpine:latest'
    name: alpine
    command:
    - echo
    - helloworld
---
apiVersion: v1
kind: Pod
metadata:
  name: alpine2
spec:
  containers:
  - image: 'alpine:latest'
    name: alpine
    command:
    - echo
    - helloworld
`,
		},
	}

	inst := cosmov1alpha1.Instance{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "test-pod-inst1",
			Namespace: nsName,
		},
		Spec: cosmov1alpha1.InstanceSpec{
			Template: cosmov1alpha1.TemplateRef{
				Name: tmpl.Name,
			},
			Override: cosmov1alpha1.OverrideSpec{},
		},
	}

	Context("when creating Template resource on new cluster", func() {
		It("should do nothing", func() {
			ctx := context.Background()

			By("creating template before instance")

			err := k8sClient.Create(ctx, &tmpl)
			Expect(err).ShouldNot(HaveOccurred())

			time.Sleep(3 * time.Second)

			var createdTmpl cosmov1alpha1.Template
			Eventually(func() error {
				return k8sClient.Get(ctx, client.ObjectKey{Name: tmpl.Name}, &createdTmpl)
			}, time.Second*10).Should(Succeed())
			Ω(ObjectSnapshot(&createdTmpl)).To(MatchSnapShot())

			var pod corev1.Pod
			key := client.ObjectKey{
				Name:      instance.InstanceResourceName(inst.Name, "alpine"),
				Namespace: nsName,
			}
			err = k8sClient.Get(ctx, key, &pod)
			Expect(apierrs.IsNotFound(err)).Should(BeTrue())
		})
	})

	Context("when creating Instance resource", func() {
		It("should do instance reconcile and create child resources", func() {
			ctx := context.Background()

			err := k8sClient.Create(ctx, &inst)
			Expect(err).ShouldNot(HaveOccurred())

			By("checking if child resources is as expected in template")
			var pod corev1.Pod
			Eventually(func() error {
				key := client.ObjectKey{
					Name:      instance.InstanceResourceName(inst.Name, "alpine"),
					Namespace: nsName,
				}
				return k8sClient.Get(ctx, key, &pod)
			}, time.Second*30).Should(Succeed())
			Ω(ObjectSnapshot(&pod)).To(MatchSnapShot())

			var pod2 corev1.Pod
			Eventually(func() error {
				key := client.ObjectKey{
					Name:      instance.InstanceResourceName(inst.Name, "alpine2"),
					Namespace: nsName,
				}
				return k8sClient.Get(ctx, key, &pod2)
			}, time.Second*30).Should(Succeed())
			Ω(ObjectSnapshot(&pod2)).To(MatchSnapShot())

			By("fetching instance resource and checking if last applied resources added in instance status")

			var createdInst cosmov1alpha1.Instance
			Eventually(func() int {
				key := client.ObjectKey{
					Name:      inst.Name,
					Namespace: nsName,
				}
				err := k8sClient.Get(ctx, key, &createdInst)
				Expect(err).ShouldNot(HaveOccurred())
				return createdInst.Status.LastAppliedObjectsCount
			}, time.Second*60).Should(BeEquivalentTo(2))
			Ω(InstanceSnapshot(&createdInst)).To(MatchSnapShot())
		})
	})

	Context("when updating Template resource", func() {
		It("should do instance reconcile and update child resources", func() {
			ctx := context.Background()

			var curInst cosmov1alpha1.Instance
			Eventually(func() error {
				key := client.ObjectKey{
					Name:      inst.Name,
					Namespace: nsName,
				}
				return k8sClient.Get(ctx, key, &curInst)
			}, time.Second*10).Should(Succeed())

			// fetch current template
			var updatedTmpl cosmov1alpha1.Template
			Eventually(func() error {
				key := types.NamespacedName{
					Name: tmpl.Name,
				}
				err := k8sClient.Get(ctx, key, &updatedTmpl)
				Expect(err).ShouldNot(HaveOccurred())

				updatedTmpl.Spec.RawYaml = `apiVersion: v1
kind: Pod
metadata:
  name: alpine
spec:
  containers:
  - image: 'alpine:next'
    name: alpine
    command:
    - echo
    - helloworld
`

				return k8sClient.Update(ctx, &updatedTmpl)
			}, time.Second*30).Should(Succeed())

			By("checking if pod updated")
			var pod corev1.Pod
			Eventually(func() string {
				key := client.ObjectKey{
					Name:      instance.InstanceResourceName(inst.Name, "alpine"),
					Namespace: nsName,
				}
				err := k8sClient.Get(ctx, key, &pod)
				Expect(err).NotTo(HaveOccurred())
				return pod.Spec.Containers[0].Image
			}, time.Second*30).Should(BeEquivalentTo("alpine:next"))
			Ω(ObjectSnapshot(&pod)).To(MatchSnapShot())

			By("checking if pod2 is deleted")
			var pod2 corev1.Pod
			Eventually(func() bool {
				key := client.ObjectKey{
					Name:      instance.InstanceResourceName(inst.Name, "alpine2"),
					Namespace: nsName,
				}
				err := k8sClient.Get(ctx, key, &pod2)
				if err != nil {
					return apierrs.IsNotFound(err)
				}
				return pod2.DeletionTimestamp != nil
			}, time.Second*60).Should(BeTrue())

			var updatedInst cosmov1alpha1.Instance
			Eventually(func() int {
				key := client.ObjectKey{
					Name:      inst.Name,
					Namespace: nsName,
				}
				err := k8sClient.Get(ctx, key, &updatedInst)
				Expect(err).ShouldNot(HaveOccurred())
				return updatedInst.Status.LastAppliedObjectsCount
			}, time.Second*60).Should(Equal(1))
			Ω(InstanceSnapshot(&updatedInst)).To(MatchSnapShot())
		})
	})
})
