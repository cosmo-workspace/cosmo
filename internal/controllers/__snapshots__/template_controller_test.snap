['Template controller when creating Instance resource should do instance reconcile and create child resources 1']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-inst1-alpine\",
    \"namespace\": \"default\",
    \"creationTimestamp\": null,
    \"labels\": {
      \"cosmo-workspace.github.io/instance\": \"test-pod-inst1\",
      \"cosmo-workspace.github.io/template\": \"test-pod-tmpl1\"
    },
    \"ownerReferences\": [
      {
        \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
        \"kind\": \"Instance\",
        \"name\": \"test-pod-inst1\",
        \"uid\": \"\",
        \"controller\": true,
        \"blockOwnerDeletion\": true
      }
    ]
  },
  \"spec\": {
    \"containers\": [
      {
        \"name\": \"alpine\",
        \"image\": \"alpine:latest\",
        \"command\": [
          \"echo\",
          \"helloworld\"
        ],
        \"resources\": {},
        \"terminationMessagePath\": \"/dev/termination-log\",
        \"terminationMessagePolicy\": \"File\",
        \"imagePullPolicy\": \"Always\"
      }
    ],
    \"restartPolicy\": \"Always\",
    \"terminationGracePeriodSeconds\": 30,
    \"dnsPolicy\": \"ClusterFirst\",
    \"securityContext\": {},
    \"schedulerName\": \"default-scheduler\",
    \"tolerations\": [
      {
        \"key\": \"node.kubernetes.io/not-ready\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      },
      {
        \"key\": \"node.kubernetes.io/unreachable\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      }
    ],
    \"priority\": 0,
    \"enableServiceLinks\": true,
    \"preemptionPolicy\": \"PreemptLowerPriority\"
  },
  \"status\": {
    \"phase\": \"Pending\",
    \"qosClass\": \"BestEffort\"
  }
}
"""

['Template controller when creating Instance resource should do instance reconcile and create child resources 2']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-inst1-alpine2\",
    \"namespace\": \"default\",
    \"creationTimestamp\": null,
    \"labels\": {
      \"cosmo-workspace.github.io/instance\": \"test-pod-inst1\",
      \"cosmo-workspace.github.io/template\": \"test-pod-tmpl1\"
    },
    \"ownerReferences\": [
      {
        \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
        \"kind\": \"Instance\",
        \"name\": \"test-pod-inst1\",
        \"uid\": \"\",
        \"controller\": true,
        \"blockOwnerDeletion\": true
      }
    ]
  },
  \"spec\": {
    \"containers\": [
      {
        \"name\": \"alpine\",
        \"image\": \"alpine:latest\",
        \"command\": [
          \"echo\",
          \"helloworld\"
        ],
        \"resources\": {},
        \"terminationMessagePath\": \"/dev/termination-log\",
        \"terminationMessagePolicy\": \"File\",
        \"imagePullPolicy\": \"Always\"
      }
    ],
    \"restartPolicy\": \"Always\",
    \"terminationGracePeriodSeconds\": 30,
    \"dnsPolicy\": \"ClusterFirst\",
    \"securityContext\": {},
    \"schedulerName\": \"default-scheduler\",
    \"tolerations\": [
      {
        \"key\": \"node.kubernetes.io/not-ready\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      },
      {
        \"key\": \"node.kubernetes.io/unreachable\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      }
    ],
    \"priority\": 0,
    \"enableServiceLinks\": true,
    \"preemptionPolicy\": \"PreemptLowerPriority\"
  },
  \"status\": {
    \"phase\": \"Pending\",
    \"qosClass\": \"BestEffort\"
  }
}
"""

['Template controller when creating Instance resource should do instance reconcile and create child resources 3']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-inst1\",
    \"namespace\": \"default\",
    \"creationTimestamp\": null
  },
  \"spec\": {
    \"template\": {
      \"name\": \"test-pod-tmpl1\"
    },
    \"override\": {}
  },
  \"status\": {
    \"templateName\": \"test-pod-tmpl1\",
    \"lastApplied\": [
      {
        \"kind\": \"Pod\",
        \"namespace\": \"default\",
        \"name\": \"test-pod-inst1-alpine\",
        \"apiVersion\": \"v1\"
      },
      {
        \"kind\": \"Pod\",
        \"namespace\": \"default\",
        \"name\": \"test-pod-inst1-alpine2\",
        \"apiVersion\": \"v1\"
      }
    ],
    \"lastAppliedObjectsCount\": 2,
    \"templateObjectsCount\": 2
  }
}
"""

['Template controller when creating Template resource on new cluster should do nothing 1']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-tmpl1\",
    \"creationTimestamp\": null
  },
  \"spec\": {
    \"rawYaml\": \"apiVersion: v1\\nkind: Pod\\nmetadata:\\n  name: alpine\\nspec:\\n  containers:\\n  - image: 'alpine:latest'\\n    name: alpine\\n    command:\\n    - echo\\n    - helloworld\\n---\\napiVersion: v1\\nkind: Pod\\nmetadata:\\n  name: alpine2\\nspec:\\n  containers:\\n  - image: 'alpine:latest'\\n    name: alpine\\n    command:\\n    - echo\\n    - helloworld\\n\"
  }
}
"""

['Template controller when updating Template resource should do instance reconcile and update child resources 1']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-inst1-alpine\",
    \"namespace\": \"default\",
    \"creationTimestamp\": null,
    \"labels\": {
      \"cosmo-workspace.github.io/instance\": \"test-pod-inst1\",
      \"cosmo-workspace.github.io/template\": \"test-pod-tmpl1\"
    },
    \"ownerReferences\": [
      {
        \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
        \"kind\": \"Instance\",
        \"name\": \"test-pod-inst1\",
        \"uid\": \"\",
        \"controller\": true,
        \"blockOwnerDeletion\": true
      }
    ]
  },
  \"spec\": {
    \"containers\": [
      {
        \"name\": \"alpine\",
        \"image\": \"alpine:next\",
        \"command\": [
          \"echo\",
          \"helloworld\"
        ],
        \"resources\": {},
        \"terminationMessagePath\": \"/dev/termination-log\",
        \"terminationMessagePolicy\": \"File\",
        \"imagePullPolicy\": \"Always\"
      }
    ],
    \"restartPolicy\": \"Always\",
    \"terminationGracePeriodSeconds\": 30,
    \"dnsPolicy\": \"ClusterFirst\",
    \"securityContext\": {},
    \"schedulerName\": \"default-scheduler\",
    \"tolerations\": [
      {
        \"key\": \"node.kubernetes.io/not-ready\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      },
      {
        \"key\": \"node.kubernetes.io/unreachable\",
        \"operator\": \"Exists\",
        \"effect\": \"NoExecute\",
        \"tolerationSeconds\": 300
      }
    ],
    \"priority\": 0,
    \"enableServiceLinks\": true,
    \"preemptionPolicy\": \"PreemptLowerPriority\"
  },
  \"status\": {
    \"phase\": \"Pending\",
    \"qosClass\": \"BestEffort\"
  }
}
"""

['Template controller when updating Template resource should do instance reconcile and update child resources 2']
SnapShot = """
{
  \"metadata\": {
    \"name\": \"test-pod-inst1\",
    \"namespace\": \"default\",
    \"creationTimestamp\": null
  },
  \"spec\": {
    \"template\": {
      \"name\": \"test-pod-tmpl1\"
    },
    \"override\": {}
  },
  \"status\": {
    \"templateName\": \"test-pod-tmpl1\",
    \"lastApplied\": [
      {
        \"kind\": \"Pod\",
        \"namespace\": \"default\",
        \"name\": \"test-pod-inst1-alpine\",
        \"apiVersion\": \"v1\"
      }
    ],
    \"lastAppliedObjectsCount\": 1,
    \"templateObjectsCount\": 1
  }
}
"""
